types/lodash/common/object.d.ts

```diff
@@ -1021,8 +1021,41 @@ declare module "../index" {
         functionsIn(): CollectionChain<string>;
     }
 
-    type GetIndexedField<T, K> = K extends keyof T
-        ? T[K]
+    /**
+     * If a value of type T can be indexed by a key of type K
+     * GetIndexedField<T, K> deduces the type of value at that key,
+     * otherwise it returns undefined type.
+     */
+    type GetIndexedField<T, K> =
+        // This covers a lot of ground: indexing arrays and strings by number,
+        // retrieving their "length", indexing objects by their proper keys.
+        K extends keyof T ? T[K]
+
+        // Corner-case: T = Record<string, SomeType>; K = number.
+        // This case does not match the condition above because K
+        // does not extend keyof T (which equals string). However,
+        // in this case it is legal to index a value of type T by
+        // a key of type K, and even TS evaluates T[K] as SomeType,
+        // accounting for JS auto-casting numeric keys to strings
+        // when indexing objects. Thus, this case gets a special
+        // treatment here.
+        : K extends number ? (`${K}` extends keyof T ? T[`${K}`] : undefined)
+
+        // TODO: This is a better variant for the rest of this function,
+        // but it needs TypeScript >= 4.8, and DefinitelyTyped still demands
```

Comment:
FYI: 4.7 support will be dropped once 5.5 is released, which should be soon according to the TS team. So one option could be to wait for that and submit the better solution.